File: Computing/comp_code/leetcode/0004_MedianOfTwoSortedArrays.py
"""
Leetcode 0004. Meadian of Two Sorted Arrays

Question:
There are two sorted arrays `nums1` and `nums2` of size `m` and `n` respectivly, return the median of the two sorted arrays. (You may assume the arrays cannot both be empty.)
Aim: 
Overall runtime complexity should be O(log(m+n)).
"""

import math

class Solution(object):
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:

        def merge(a, b):
            """merges two already sorted arrays"""
            a.append(math.inf)
            b.append(math.inf)
            c = []

            for i in range(len(a)+len(b)-2):
                if a[0] <= b[0]:
                    c.append(a.pop(0))
                else:
                    c.append(b.pop(0))
            return c

        nums3 = merge(nums1, nums2)

        if len(nums3) % 2 == 0:
            return ((nums3[int(len(nums3)/2)-1] + nums3[int(len(nums3)/2)])/2)
        else:
            return nums3[math.floor(len(nums3)/2.0)]

"""
Submission:
88ms, faster than 80.44%
14.2Mb, less than 13%

Complexity Anaylsis:
- O(n)

Notes:
- can be done faster, namely O(log(min(m,n)))
"""

File: Computing/comp_code/leetcode/0004_MedianOfTwoSortedArrays.hs
{- Leetcode 0004. Meadian of Two Sorted Arrays

Question:
There are two sorted arrays `nums1` and `nums2` of size `m` and `n` respectivly, return the median of the two sorted arrays. (You may assume the arrays cannot both be empty.)
Aim: 
Overall runtime complexity should be O(log(m+n)).
-}

module LeetCode0004MedianOfTwoSortedArrays (medianOfTwoSortedArrrays) where

import Test.QuickCheck

medianOfTwoSortedArrrays :: (Ord a, Integral a, Fractional b) => [a] -> [a] -> b
medianOfTwoSortedArrrays xs ys = median $ map fromIntegral $ merge xs ys

merge :: Ord a => [a] -> [a] -> [a]
merge [] ys = ys
merge xs [] = xs
merge (x:xs) (y:ys) | x <= y    = x : merge xs (y:ys)
                    | otherwise = y : merge (x:xs) ys

median :: (Fractional a) => [a] -> a
median xs | odd  len = xs !! mid 
          | even len = (xs !! (mid - 1) + xs !! mid) / 2 
  where len = length xs
        mid = len `div` 2 


-- Testing
prop_01 = medianOfTwoSortedArrrays [1,3] [2] == 2
prop_02 = medianOfTwoSortedArrrays [1,2] [3,4] == 2.5
prop_03 = medianOfTwoSortedArrrays [0,0] [0,0] == 0
prop_04 = medianOfTwoSortedArrrays [] [1] == 1


{-
Complexity Anaylsis:
- O(n)

Notes:
- can be done faster, namely O(log(min(m,n)))
-}
